components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -494,6 +499,33 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         return user;
     }
 
+    public async supportConfigureWorkspaceTimeout(ctx: TraceContext): Promise<boolean> {
+        const user = this.checkUser("supportConfigureWorkspaceTimeout");
+        await this.guardAccess({ kind: "user", subject: user }, "get");
+
+        return await this.entitlementService.maySetTimeout(user, new Date());
+    }
+
+    public async updateWorkspaceTimeoutSetting(
+        ctx: TraceContext,
+        setting: Partial<WorkspaceTimeoutSetting>,
+    ): Promise<void> {
+        traceAPIParams(ctx, { setting }); // partialUser contains PII
+
+        const user = this.checkUser("updateWorkspaceTimeoutSetting");
+        await this.guardAccess({ kind: "user", subject: user }, "update");
+
+        if (!(await this.entitlementService.maySetTimeout(user, new Date()))) {
+            throw new Error("configure workspace timeout only available for paid user.");
+        }
+        if (setting.workspaceTimeout) {
+            WorkspaceTimeoutDuration.validate(setting.workspaceTimeout);
+        }
```

Comment:
You can shift this to happen earlier, even before checking authorization. Being able to reject invalid requests early is valuable in ensuring we don't do too much work, but also guard against amplification attacks.